powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/v4/kdf/Kdf.java

```diff
@@ -0,0 +1,95 @@
+ * limitations under the License.
+ */
+
+package io.getlime.security.powerauth.crypto.lib.v4.kdf;
+
+import io.getlime.security.powerauth.crypto.lib.model.exception.GenericCryptoException;
+import io.getlime.security.powerauth.crypto.lib.util.ByteUtils;
+import io.getlime.security.powerauth.crypto.lib.util.KeyConvertor;
+import org.bouncycastle.crypto.macs.KMAC;
+import org.bouncycastle.crypto.params.KeyParameter;
+
+import javax.crypto.SecretKey;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Universal KDF based on KMAC-256 (Keccak).
+ *
+ * @author Roman Strobl, roman.strobl@wultra.com
+ */
+public class Kdf {
+
+    private static final String CRYPTO4_KDF_CUSTOM_STRING = "PA4KDF";
+    private static final Integer KMAC_ALGORITHM_STRENGTH = 256;
+
+    private static final KeyConvertor KEY_CONVERTOR = new KeyConvertor();
+
+    /**
+     * Derive a secret key based on an input key, numeric key index, requested key size and optional context.
+     * @param key Secret key to be used for key derivation.
+     * @param index Key index (numeric).
+     * @param outLength Requested derived key size.
+     * @param context Optional context to use during key derivation.
+     * @return Derived secret key.
+     * @throws GenericCryptoException Thrown in case of any cryptography error.
+     */
+    public static SecretKey derive(SecretKey key, long index, int outLength, byte[] context) throws GenericCryptoException {
+        final byte[] indexBytes = ByteUtils.encodeLong(index);
+        final byte[] data;
+        if (context != null) {
+            data = ByteUtils.concat(indexBytes, ByteUtils.concatWithSizes(context));
+        } else {
+            data = indexBytes;
+        }
+        final byte[] output = kmac256(key, data, outLength, CRYPTO4_KDF_CUSTOM_STRING.getBytes(StandardCharsets.UTF_8));
+        return KEY_CONVERTOR.convertBytesToSharedSecretKey(output);
+    }
+
+    /**
+     * Compute the KMAC256 of the given data using provided secret key, output length and optional customization string.
+     *
+     * @param key          The secret key, must be a valid {@link SecretKey} with a 256-bit key length.
+     * @param data         The input data used for the KMAC.
+     * @param outLength    The length of generated output bytes.
+     * @param customString An optional customization string, use null value for no customization.
+     * @return KMAC256 output byte array.
+     * @throws GenericCryptoException Thrown in case of any cryptography error.
+     */
+    static byte[] kmac256(SecretKey key, byte[] data, int outLength, byte[] customString) throws GenericCryptoException {
+        // Validate inputs
+        if (key == null) {
+            throw new GenericCryptoException("Missing secret key for KDF.");
+        }
+        if (data == null) {
+            throw new GenericCryptoException("Missing data for KDF.");
+        }
+        if (outLength < 32) {
```

Comment:
Should we maybe check if this is some multiple, i.e., modulo some block size?

Reply:
I am not sure how to do that properly. If you dislike the magic constant which is inline with use cases of the crypto library, I can check that the outputLenght is greater than 0? @hvge what do you think?

Reply:
According to NIST SP-800-185:

> When used as a MAC, applications of this Recommendation shall not select an output length L
that is less than 32 bits, and shall only select an output length less than 64 bits after a careful risk
analysis is performed.

In my design, we use 16 bytes (128 bits) for PowerAuth online authorization code components ("signature" in our old naming). We can extend this of course, but I would keep recommendation from NIST and fail if length is less than 8B.

It's complete legit to use output length 33, or more.